### PR TITLE
Change RTL detection to rely on unicode-bidi paragraph by paragraph

### PR DESCRIPTION
**Pull request by Gargron**
Ref: https://discourse.joinmastodon.org/t/improve-bidi-support-a-better-support-for-rtl/2876

### CODE REVIEW

File: app/javascript/styles/mastodon/components.scss [hunk 1]
```diff
@@ -821,6 +822,7 @@
   a {
     color: $secondary-text-color;
     text-decoration: none;
+    unicode-bidi: isolate;
```
Comment (ahangarha) [inline app/javascript/styles/mastodon/components.scss:825]: I think this shouldn't be applied for all links but links such as mentions and hashtags.

Comment (Gargron) [reply app/javascript/styles/mastodon/components.scss:825]: But all links are ascii (`https://...`)

Comment (ThibG) [reply app/javascript/styles/mastodon/components.scss:825]: Not necessarily. Mastodon allows links with arbitrary text from other software.

Comment (Gargron) [reply app/javascript/styles/mastodon/components.scss:825]: I would wager that the *majority* of links would be ascii URLs and we'll have consistently better results by not making them affect the directionality of the whole post.

**Comment by ThibG** [comment]
It might not be well-supported in Safari: https://caniuse.com/#feat=mdn-css_properties_unicode-bidi_plaintext

**Comment by Gargron** [comment]
Okay, so I see two options here. Maybe we can get away with this not working in Safari, who knows how many Safari users we have that would be affected by this (directionality of the posts would be affected by interface language, so if you use the UI in a right-to-left language, you'd see posts right-to-left as well). Or maybe we keep the flawed detection code and the `dir` attributes and hopefully the `unicode-bidi` property overrides that when the browser supports it.

**Comment by ThibG** [comment]
> Okay, so I see two options here. Maybe we can get away with this not working in Safari, who knows how many Safari users we have that would be affected by this (directionality of the posts would be affected by interface language, so if you use the UI in a right-to-left language, you'd see posts right-to-left as well).

It could be ok, but I'm not sure…

> Or maybe we keep the flawed detection code and the `dir` attributes and hopefully the `unicode-bidi` property overrides that when the browser supports it.

If I understand the specs correctly, the `plaintext` and `isolate` values would indeed override the `dir` attributes where supported.

That being said, the spec says “Authors should not use unicode-bidi in HTML documents.”

An alternative could be to *not* set the direction on toots, and wrap them and each of their `<p>` in a `<bdi>`

**Comment by ahangarha** [comment]
I have noticed in some cases, still there are some issue with using unicode--bidi. It is very tricky when the text is mixed of LTR and RTL characters.

If you can change the structure of toots by adding `dir="auto"` to `p` and then use some CSS tweaks, I think it would be much much easier to handle this issue.

File: app/javascript/styles/mastodon/components.scss (hunk 1 above)
Comment (ahangarha) [reply app/javascript/styles/mastodon/components.scss:825]: Is there any way to bring a live instance with these changes so I test multiple possible scenarios?

**Comment by Gargron** [comment]
I have heard that `dir="auto"` takes care of detecting text direction in input elements. So, `rtl.js` can be fully removed.

I'm in favour of relying on browser support here.

**Comment by ahangarha** [comment]
Using 'dir=auto' will handle everything but excluding mentions in the beginning of the paragraph.

**Comment by Gargron** [comment]
My motivation here is that #15218 seems to indicate some kind of recursion error in the `isRtl` JavaScript function (though I fail to see it). This PR is a one stone two birds kind of deal in that regard.

**Review by ykzts — APPROVED** [review]

**Comment by ahangarha** [comment]
Great. Counting seconds to see this change on instances.

Thank you

**Comment by ahangarha** [comment]
I have noticed a little issue when the language is set to Persian (and for sure other RTL ones) which is making little issue with paragraphs starting with LTR characters.

This can be solved by adding `direction: initial;` to `.reply-indicator__content p, .status__content p`.

**Pull request merged**